### PR TITLE
refactor: use class for action result

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

File: mergify_engine/actions/__init__.py [hunk 1]
```diff
@@ -46,6 +47,32 @@ def get_commands():
     return dict((name, obj) for name, obj in get_classes().items() if obj.is_command)
 
 
+@dataclasses.dataclass  # type: ignore
```
Comment (jd) [inline mergify_engine/actions/__init__.py:50]: sounds like it's not needed here since it's an enum

File: mergify_engine/actions/__init__.py [hunk 2]
```diff
@@ -46,6 +47,32 @@ def get_commands():
...
+    PENDING = None
+    CANCELLED = "cancelled"
+    SUCCESS = "success"
+    FAILURE = "failure"
+    NEUTRAL = "neutral"
+    ACTION_REQUIRED = "action_required"
+
+    def get_status(self):
+        if self == Conclusion.PENDING:
+            return "in_progress"
```
Comment (jd) [inline mergify_engine/actions/__init__.py:61]: Can we have an `Enum` for that too?

File: mergify_engine/actions/__init__.py [hunk 3]
```diff
@@ -46,6 +47,32 @@ def get_commands():
...
+            return "completed"
+
+
+@dataclasses.dataclass  # type: ignore
+class Result:
+    conclusion: Conclusion
+    title: str
+    summary: str
+
+    def __repr__(self):
```
Comment (jd) [inline mergify_engine/actions/__init__.py:72]: this is generated by `dataclass`

File: mergify_engine/actions/__init__.py [hunk 4]
```diff
@@ -46,6 +47,32 @@ def get_commands():
...
+    ACTION_REQUIRED = "action_required"
+
+    def get_status(self):
+        if self == Conclusion.PENDING:
+            return "in_progress"
+        else:
+            return "completed"
+
+
+@dataclasses.dataclass  # type: ignore
```
Comment (jd) [inline mergify_engine/actions/__init__.py:66]:
```suggestion
@dataclasses.dataclass
```

File: mergify_engine/actions/assign.py [hunk 5]
```diff
@@ -51,10 +53,14 @@ def run(self, ctxt, rule, missing_conditions):
...
-                return (
-                    None,
+                return actions.Result(
+                    actions.Conclusion.PENDING,
                     "Unable to add assignees",
                     f"GitHub error: [{e.status_code}] `{e.message}`",
                 )
 
-        return ("success", "Assignees added", ", ".join(self.config["users"]))
+        return actions.Result(
```
Comment (jd) [inline mergify_engine/actions/assign.py:62]: you might want to add type hint to the return type of the method

File: mergify_engine/actions/merge/queue.py [hunk 6]
```diff
@@ -188,20 +189,19 @@ def handle_first_pull_in_queue(self, ctxt):
...
+        result = helpers.merge_report(ctxt, True)
+        if result:
             ctxt.log.info(
                 "pull request closed in the meantime",
-                conclusion=conclusion,
-                title=title,
-                summary=summary,
+                conclusion=result.conclusion,
+                title=result.title,
+                summary=result.summary,
```
Comment (jd) [inline mergify_engine/actions/merge/queue.py:198]: could do `result=result` though

**Review by jd — DISMISSED** [review]

File: mergify_engine/actions/merge/queue.py (hunk 6 above)
Comment (sileht) [reply mergify_engine/actions/merge/queue.py:197]: You read in my mind ;)

Comment (jd) [reply mergify_engine/actions/merge/queue.py:197]: Not really I guess?

File: mergify_engine/actions/post_check.py [hunk 7]
```diff
@@ -86,13 +87,16 @@ def _post(self, ctxt, rule, missing_conditions):
...
+                check_api.Conclusion.FAILURE,
                 "Invalid summary template",
                 str(rmf),
             )
 
-        return ("failure" if missing_conditions else "success", title, summary)
+        if missing_conditions:
+            return check_api.Result(check_api.Conclusion.FAILURE, title, summary)
+        else:
+            return check_api.Result(check_api.Conclusion.SUCCESS, title, summary)
```
Comment (jd) [inline mergify_engine/actions/post_check.py:99]: This seems to be outside the scope of this patch?

File: mergify_engine/tests/functional/base.py [hunk 8]
```diff
@@ -553,6 +553,8 @@ def tearDown(self):
 
         loop = asyncio.get_event_loop()
         loop.run_until_complete(web.shutdown())
+        for task in asyncio.all_tasks(loop):
+            task.cancel()
```
Comment (jd) [inline mergify_engine/tests/functional/base.py:557]: looks unrelated :(

**Review by jd — CHANGES_REQUESTED** [review]
A couple of changes seems unrelated though
Otherwise it looks beautiful 👍

**Comment by jd** [comment]
My question around logging is, can we just do `log(result=result` rather than logging each field individually?

File: mergify_engine/actions/merge/queue.py (hunk 6 above)
Comment (sileht) [reply mergify_engine/actions/merge/queue.py:197]: Oh I change `set_check_run` and `set_summary_check` signature to take `result` directly. But I didn't do it for logging to keep these attributes searcheable in datadog.

File: mergify_engine/actions/post_check.py (hunk 7 above)
Comment (sileht) [reply mergify_engine/actions/post_check.py:99]: Hum, it's the same code as before with the new enum/dataclass

**Comment by sileht** [comment]
> My question around logging is, can we just do `log(result=result)` rather than logging each field individually?

I wanted to keep it searcheable in datadog, but after some though this morning, I never ever use these attributes to filter logs, so.

File: mergify_engine/actions/post_check.py (hunk 7 above)
Comment (jd) [reply mergify_engine/actions/post_check.py:101]: My bad I misread the original code with the hidden `if`

**Review by jd — APPROVED** [review]
